```
FAIL	compress/zlib	0.000s
# internal/syscall/unix
/Users/ayke/lib/go/src/internal/syscall/unix/net_darwin.go:34:21: undefined: syscall.RawSockaddr
/Users/ayke/lib/go/src/internal/syscall/unix/net.go:16:63: undefined: syscall.SockaddrInet4
/Users/ayke/lib/go/src/internal/syscall/unix/net.go:20:63: undefined: syscall.SockaddrInet6
/Users/ayke/lib/go/src/internal/syscall/unix/net.go:24:59: undefined: syscall.SockaddrInet4
/Users/ayke/lib/go/src/internal/syscall/unix/net.go:28:59: undefined: syscall.SockaddrInet6
/Users/ayke/lib/go/src/internal/syscall/unix/net.go:32:55: undefined: syscall.SockaddrInet4
/Users/ayke/lib/go/src/internal/syscall/unix/net.go:36:55: undefined: syscall.SockaddrInet6
/Users/ayke/lib/go/src/internal/syscall/unix/net.go:40:67: undefined: syscall.SockaddrInet4
/Users/ayke/lib/go/src/internal/syscall/unix/net.go:44:67: undefined: syscall.SockaddrInet6
/Users/ayke/lib/go/src/internal/syscall/unix/net_darwin.go:70:30: undefined: syscall.RawSockaddr
/Users/ayke/lib/go/src/internal/syscall/unix/nonblocking_libc.go:15:32: undefined: syscall.F_GETFL
/Users/ayke/lib/go/src/internal/syscall/unix/nonblocking_libc.go:19:22: undefined: syscall.O_NONBLOCK
```

This one is going to be fun.

--- a/.circleci/config.yml
+++ b/.circleci/config.yml
@@ -105,13 +105,6 @@ jobs:
       - test-linux:
           llvm: "14"
     resource_class: large
-  test-llvm15-go120:
-    docker:
-      - image: golang:1.20-rc-buster
-    steps:
-      - test-linux:
-          llvm: "15"
-    resource_class: large
 
 workflows:
   test-all:
@@ -119,4 +112,3 @@ workflows:
       # This tests our lowest supported versions of Go and LLVM, to make sure at
       # least the smoke tests still pass.
       - test-llvm14-go118
-      - test-llvm15-go120

--- a/.github/workflows/build-macos.yml
+++ b/.github/workflows/build-macos.yml
@@ -27,7 +27,7 @@ jobs:
       - name: Install Go
         uses: actions/setup-go@v3
         with:
-          go-version: '1.19'
+          go-version: '1.20'
           cache: true
       - name: Cache LLVM source
         uses: actions/cache@v3
@@ -107,7 +107,7 @@ jobs:
       - name: Install Go
         uses: actions/setup-go@v3
         with:
-          go-version: '1.19'
+          go-version: '1.20'
           cache: true
       - name: Build TinyGo
         run: go install

--- a/.github/workflows/linux.yml
+++ b/.github/workflows/linux.yml
@@ -18,7 +18,7 @@ jobs:
     # statically linked binary.
     runs-on: ubuntu-latest
     container:
-      image: golang:1.19-alpine
+      image: golang:1.20-alpine
     steps:
       - name: Install apk dependencies
         # tar: needed for actions/cache@v3
@@ -118,7 +118,7 @@ jobs:
       - name: Install Go
         uses: actions/setup-go@v3
         with:
-          go-version: '1.19'
+          go-version: '1.20'
           cache: true
       - name: Install wasmtime
         run: |
@@ -158,7 +158,7 @@ jobs:
       - name: Install Go
         uses: actions/setup-go@v3
         with:
-          go-version: '1.19'
+          go-version: '1.20'
           cache: true
       - name: Install Node.js
         uses: actions/setup-node@v3
@@ -252,7 +252,7 @@ jobs:
       - name: Install Go
         uses: actions/setup-go@v3
         with:
-          go-version: '1.19'
+          go-version: '1.20'
           cache: true
       - name: Cache LLVM source
         uses: actions/cache@v3
@@ -352,7 +352,7 @@ jobs:
       - name: Install Go
         uses: actions/setup-go@v3
         with:
-          go-version: '1.19'
+          go-version: '1.20'
           cache: true
       - name: Cache LLVM source
         uses: actions/cache@v3

--- a/.github/workflows/windows.yml
+++ b/.github/workflows/windows.yml
@@ -29,7 +29,7 @@ jobs:
       - name: Install Go
         uses: actions/setup-go@v3
         with:
-          go-version: '1.19'
+          go-version: '1.20'
           cache: true
       - name: Cache LLVM source
         uses: actions/cache@v3
@@ -112,7 +112,7 @@ jobs:
       - name: Install Go
         uses: actions/setup-go@v3
         with:
-          go-version: '1.19'
+          go-version: '1.20'
           cache: true
       - name: Download TinyGo build
         uses: actions/download-artifact@v2
@@ -136,7 +136,7 @@ jobs:
       - name: Install Go
         uses: actions/setup-go@v3
         with:
-          go-version: '1.19'
+          go-version: '1.20'
           cache: true
       - name: Download TinyGo build
         uses: actions/download-artifact@v2
@@ -166,7 +166,7 @@ jobs:
       - name: Install Go
         uses: actions/setup-go@v3
         with:
-          go-version: '1.19'
+          go-version: '1.20'
           cache: true
       - name: Download TinyGo build
         uses: actions/download-artifact@v2

--- a/Dockerfile
+++ b/Dockerfile
@@ -1,5 +1,5 @@
 # tinygo-llvm stage obtains the llvm source for TinyGo
-FROM golang:1.19 AS tinygo-llvm
+FROM golang:1.20 AS tinygo-llvm
 
 RUN apt-get update && \
     apt-get install -y apt-utils make cmake clang-11 ninja-build

--- a/src/runtime/env.go
+++ b/src/runtime/env.go
@@ -1,0 +1,51 @@
+//go:build linux || darwin || windows
+
+package runtime
+
+// Update the C environment if cgo is loaded.
+// Called from Go 1.20 and above.
+//
+//go:linkname syscallSetenv syscall.runtimeSetenv
+func syscallSetenv(key, value string) {
+	keydata := cstring(key)
+	valdata := cstring(value)
+	setenv(&keydata[0], &valdata[0])
+	if key == "GODEBUG" && godebugUpdate != nil {
+		// Starting with Go 1.20, we need to call a callback (set by
+		// internal/godebug) to notify the GODEBUG environment variable has
+		// changed. This is necessary to get archive/zip to pass tests.
+		godebugUpdate(key, value)
+	}
+}
+
+// Update the C environment if cgo is loaded.
+// Called from Go 1.20 and above.
+//
+//go:linkname syscallUnsetenv syscall.runtimeUnsetenv
+func syscallUnsetenv(key string) {
+	keydata := cstring(key)
+	unsetenv(&keydata[0])
+}
+
+// Compatibility with Go 1.19 and below.
+//
+//go:linkname syscall_setenv_c syscall.setenv_c
+func syscall_setenv_c(key string, val string) {
+	syscallSetenv(key, val)
+}
+
+// Compatibility with Go 1.19 and below.
+//
+//go:linkname syscall_unsetenv_c syscall.unsetenv_c
+func syscall_unsetenv_c(key string) {
+	syscallUnsetenv(key)
+}
+
+// cstring converts a Go string to a C string.
+// borrowed from syscall
+func cstring(s string) []byte {
+	data := make([]byte, len(s)+1)
+	copy(data, s)
+	// final byte should be zero from the initial allocation
+	return data
+}

--- a/src/runtime/env_unix.go
+++ b/src/runtime/env_unix.go
@@ -2,56 +2,6 @@
 
 package runtime
 
-// Update the C environment if cgo is loaded.
-// Called from Go 1.20 and above.
-//
-//go:linkname syscallSetenv syscall.runtimeSetenv
-func syscallSetenv(key, value string) {
-	keydata := cstring(key)
-	valdata := cstring(value)
-	// ignore any errors
-	libc_setenv(&keydata[0], &valdata[0], 1)
-	if key == "GODEBUG" && godebugUpdate != nil {
-		// Starting with Go 1.20, we need to call a callback (set by
-		// internal/godebug) to notify the GODEBUG environment variable has
-		// changed. This is necessary to get archive/zip to pass tests.
-		godebugUpdate(key, value)
-	}
-}
-
-// Update the C environment if cgo is loaded.
-// Called from Go 1.20 and above.
-//
-//go:linkname syscallUnsetenv syscall.runtimeUnsetenv
-func syscallUnsetenv(key string) {
-	keydata := cstring(key)
-	// ignore any errors
-	libc_unsetenv(&keydata[0])
-}
-
-// Compatibility with Go 1.19 and below.
-//
-//go:linkname syscall_setenv_c syscall.setenv_c
-func syscall_setenv_c(key string, val string) {
-	syscallSetenv(key, val)
-}
-
-// Compatibility with Go 1.19 and below.
-//
-//go:linkname syscall_unsetenv_c syscall.unsetenv_c
-func syscall_unsetenv_c(key string) {
-	syscallUnsetenv(key)
-}
-
-// cstring converts a Go string to a C string.
-// borrowed from syscall
-func cstring(s string) []byte {
-	data := make([]byte, len(s)+1)
-	copy(data, s)
-	// final byte should be zero from the initial allocation
-	return data
-}
-
 // int setenv(const char *name, const char *val, int replace);
 //
 //export setenv
@@ -61,3 +11,13 @@ func libc_setenv(name *byte, val *byte, replace int32) int32
 //
 //export unsetenv
 func libc_unsetenv(name *byte) int32
+
+func setenv(key, val *byte) {
+	// ignore any errors
+	libc_setenv(key, val, 1)
+}
+
+func unsetenv(key *byte) {
+	// ignore any errors
+	libc_unsetenv(key)
+}

--- a/src/runtime/env_unix.go
+++ b/src/runtime/env_unix.go
@@ -11,6 +11,12 @@ func syscallSetenv(key, value string) {
 	valdata := cstring(value)
 	// ignore any errors
 	libc_setenv(&keydata[0], &valdata[0], 1)
+	if key == "GODEBUG" && godebugUpdate != nil {
+		// Starting with Go 1.20, we need to call a callback (set by
+		// internal/godebug) to notify the GODEBUG environment variable has
+		// changed. This is necessary to get archive/zip to pass tests.
+		godebugUpdate(key, value)
+	}
 }
 
 // Update the C environment if cgo is loaded.

--- a/src/runtime/env_windows.go
+++ b/src/runtime/env_windows.go
@@ -1,0 +1,23 @@
+//go:build windows
+
+package runtime
+
+// Set environment variable in Windows:
+//
+//	BOOL SetEnvironmentVariableA(
+//	    [in]           LPCSTR lpName,
+//	    [in, optional] LPCSTR lpValue
+//	);
+//
+//export SetEnvironmentVariableA
+func _SetEnvironmentVariableA(key, val *byte) bool
+
+func setenv(key, val *byte) {
+	// ignore any errors
+	_SetEnvironmentVariableA(key, val)
+}
+
+func unsetenv(key *byte) {
+	// ignore any errors
+	_SetEnvironmentVariableA(key, nil)
+}

--- a/src/runtime/runtime.go
+++ b/src/runtime/runtime.go
@@ -96,8 +96,11 @@ func SetFinalizer(obj interface{}, finalizer interface{}) {
 	// Unimplemented.
 }
 
+var godebugUpdate func(string, string)
+
 //go:linkname godebug_setUpdate internal/godebug.setUpdate
 func godebug_setUpdate(update func(string, string)) {
-	// Unimplemented. The 'update' function needs to be called whenever the
-	// GODEBUG environment variable changes (for example, via os.Setenv).
+	// The 'update' function needs to be called whenever the GODEBUG environment
+	// variable changes (for example, via os.Setenv).
+	godebugUpdate = update
 }

